Firestore/core/src/model/mutable_document.cc

```diff
@@ -0,0 +1,137 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/model/mutable_document.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+/* static */
+MutableDocument MutableDocument::InvalidDocument(
+    const firebase::firestore::model::DocumentKey& document_key) {
+  return MutableDocument{document_key, DocumentType::kInvalid,
+                         SnapshotVersion::None(), ObjectValue{},
+                         DocumentState::kSynced};
+}
+
+/* static */
+MutableDocument MutableDocument::FoundDocument(
+    const firebase::firestore::model::DocumentKey& document_key,
+    const firebase::firestore::model::SnapshotVersion& version,
+    firebase::firestore::model::ObjectValue value) {
+  return std::move(InvalidDocument(document_key)
+                       .ConvertToFoundDocument(version, std::move(value)));
+}
+
+/* static */
+MutableDocument MutableDocument::NoDocument(
+    const firebase::firestore::model::DocumentKey& document_key,
+    const firebase::firestore::model::SnapshotVersion& version) {
+  return std::move(InvalidDocument(document_key).ConvertToNoDocument(version));
+}
+
+/* static */
+MutableDocument MutableDocument::UnknownDocument(
+    const firebase::firestore::model::DocumentKey& document_key,
+    const firebase::firestore::model::SnapshotVersion& version) {
+  return std::move(
+      InvalidDocument(document_key).ConvertToUnknownDocument(version));
+}
+
+MutableDocument& MutableDocument::ConvertToFoundDocument(
+    const firebase::firestore::model::SnapshotVersion& version,
+    firebase::firestore::model::ObjectValue value) {
+  version_ = version;
+  document_type_ = DocumentType::kFoundDocument;
+  value_ = std::move(value);
+  document_state_ = DocumentState::kSynced;
+  return *this;
+}
+
+MutableDocument& MutableDocument::ConvertToNoDocument(
+    const firebase::firestore::model::SnapshotVersion& version) {
+  version_ = version;
+  document_type_ = DocumentType::kNoDocument;
+  value_ = {};
+  document_state_ = DocumentState::kSynced;
+  return *this;
+}
+
+MutableDocument& MutableDocument::ConvertToUnknownDocument(
+    const firebase::firestore::model::SnapshotVersion& version) {
+  version_ = version;
+  document_type_ = DocumentType::kUnknownDocument;
+  value_ = {};
+  document_state_ = DocumentState::kHasCommittedMutations;
+  return *this;
+}
+
+MutableDocument& MutableDocument::SetHasCommittedMutations() {
+  document_state_ = DocumentState::kHasCommittedMutations;
+  return *this;
+}
+
+MutableDocument& MutableDocument::SetHasLocalMutations() {
+  document_state_ = DocumentState::kHasLocalMutations;
+  return *this;
+}
+
+bool operator==(const MutableDocument& lhs, const MutableDocument& rhs) {
+  return lhs.key_ == rhs.key_ && lhs.document_type_ == rhs.document_type_ &&
+         lhs.version_ == rhs.version_ && lhs.value_ == rhs.value_ &&
```

Comment:
Ultranit: value can be the last check, they can be potentially big.

Reply:
Makes sense.